### ISSUE TITLE
Pathological performance with large config file

### ISSUE THREAD

**Issue by alexcrichton** [libgit2/libgit2#3725]
Downstream bug report is here: https://github.com/rust-lang/cargo/issues/2540, but the gist of it is that a massive `.gitconfig` file can cause pathological perfomance for simple operations.

For example, this program:

```c
#include <assert.h>                                               
#include <git2.h>                                                 
                                                                  
int main() {                                                      
  assert(git_libgit2_init() == 1);                                
  git_repository *out;                                            
  assert(git_repository_open(&out, "/path/to/a/repo") == 0);
}                                                                 
```

Takes 1s locally to run when `~/.gitconfig` has 20k instances of `ui=true`. Profiling shows that [this loop](https://github.com/libgit2/libgit2/blob/381caf56908ad9cdb5959850a331d7cf2f952e63/src/config_file.c#L185) is where basically all runtime is, so maybe it's an O(n^2) thing when it needs to be O(n)?

**Comment by carlosmn** [comment]
That is indeed a naïve loop, but that's because the expected use-case is to have two or three values per multi-var, not 20k.

We could keep a pointer to the tail to make appending constant-time, but it's not a situation where I feel particularly bad being slow.

**Comment by alexcrichton** [comment]
Yeah that's what I figured :). Just seemed "responsible" to report upstream, but certainly not blocking for us either!
